### PR TITLE
Improved optimal parser with dictionary

### PR DESCRIPTION
**Pull request by Cyan4973**
This patch _slightly_ improves performance of `zstd` optimal parser
when associated with dictionary.

The main issue is that the optimal parser wasn't aware that it was processing data for dictionary compression. As a consequence, it would initialize its statistics to some "default", which are not great, and certainly not specifically adapted to dictionary. 

`opt` statistics can adapt to source over time, but dictionary compression is mostly about small data, so there is no time to adapt. On top of that, a cheap heuristics determines that, when `srcSize` is "small" (<1 KB), statistics are not even used, and some default static values are used instead.
This leads to choices which are less optimal to the real symbol distribution.
In most cases, real distribution is expected be faithful to what the dictionary predicts, and the finalized compressed block is expected to use just that.

This patch makes the optimal parser aware that it is compressing for dictionary, and able to collect its statistics, resulting in better initialization, hence better choices, "tuned" for the kind of data expected by dictionary.

Example on ["github_users"](https://github.com/facebook/zstd/releases/tag/v1.1.3) corpus : 

| level | Ratio for `dev` | Ratio for this patch | improvement |
| --- | --- | --- | --- |
| 14 | 9.276 | 9.411 | +1.45% |
| 15 | 9.460 | 9.718 | +2.73% |
| 16 | 9.439 | 9.863 | +4.49% |
| 17 | 9.940 | 10.40 | +4.68% |
| 18 | 10.38 | 10.77 | +3.84% |
| 20 | 10.38 | 10.87 | +4.72% |

It's only a few %, but it's definitely a lot for the high compression mode.

The statistics are used directly when input to compress is small, which is generally the expected scenario for dictionary compression.

Should the input be actually "large" (defined as > 8 KB), the algorithm will still use dynamically adapted statistics. But it will seed them from the dictionary, instead of starting from some generic "default".
This is better than just blindly trusting dictionary, since for large files, accurate statistics description is more important, hence even a small difference matters.

### CODE REVIEW

File: lib/compress/zstd_opt.c [hunk 1]
```diff
@@ -32,34 +32,100 @@ static void ZSTD_setLog2Prices(optState_t* optPtr)
...
+                    optPtr->offCodeSum += optPtr->offCodeFreq[of];
+            }   }
 
-    } else {
+        } else {  /* not a dictionary */
+
+            assert(optPtr->litFreq != NULL);
+            optPtr->litSum = 0;
+            {   unsigned lit = MaxLit;
+                FSE_count(optPtr->litFreq, &lit, src, srcSize);   /* use raw first block to init statistics */
```
[the remark this answers is not in the document]
Comment (Cyan4973) [reply lib/compress/zstd_opt.c]: Right, 
if the optimal parser is going to be used in kernel space,
that's going to be necessary.

File: lib/compress/zstd_opt.c [hunk 2]
```diff
@@ -69,53 +135,83 @@ static void ZSTD_rescaleFreqs(optState_t* const optPtr,
         }
         optPtr->litLengthSum = 0;
         for (u=0; u<=MaxLL; u++) {
-            optPtr->litLengthFreq[u] = 1 + (optPtr->litLengthFreq[u]>>(ZSTD_FREQ_DIV+1));
+            optPtr->litLengthFreq[u] = 1 + (optPtr->litLengthFreq[u] >> ZSTD_FREQ_DIV);
```
[the remark this answers is not in the document]
Comment (Cyan4973) [reply lib/compress/zstd_opt.c]: No idea. Maybe some hand-tuned optimization over-fitting a specific scenario.
or a copy/paste mistake.
I checked the impact of changing the update rate of this symbol to match other FSE symbols : it's generally favorable, though by a tiny amount.

File: lib/compress/zstd_opt.c [hunk 3]
```diff
@@ -32,34 +32,100 @@ static void ZSTD_setLog2Prices(optState_t* optPtr)
...
 
-    if (optPtr->litLengthSum == 0) {  /* first init */
-        unsigned u;
-        if (srcSize <= 1024) optPtr->staticPrices = 1;
+            assert(optPtr->litFreq != NULL);
+            assert(optPtr->symbolCosts != NULL);
+            optPtr->litSum = 0;
+            {   unsigned lit;
+                for (lit=0; lit<=MaxLit; lit++) {
+                    U32 const scaleLog = 11;   /* scale to 2K */
```
[the remark this answers is not in the document]
Comment (Cyan4973) [reply lib/compress/zstd_opt.c]: I tried a few values. This one ended up the better one.
Note that, as long as `scaleLog` is not stupidly large nor stupidly small, differences are small.

[the remark this answers is not in the document]
Comment (terrelln) [reply lib/compress/zstd_opt.c]: `scaleLog` changes how fast the source adapts for sources > 8KB right?

[the remark this answers is not in the document]
Comment (Cyan4973) [reply lib/compress/zstd_opt.c]: `scaleLog` changes the relative weight of initial statistics.
If too small, initial statistics will be meaningless, too close to average for all symbols.
If too strong, the dynamic part bringing "real" symbols will adapt too slowly.
It's a balancing exercise.

**Pull request merged**